cli/spec.py

```diff
@@ -16,30 +20,187 @@
+        by type.
+
+        :param spec: dict. spec file
+        :param args: dict. argparse arguments.
+        :param app_settings_dir: path to the base directory holding the
+            application's settings. App can be provisioner\installer\tester
+            and the path would be: settings/<app_name>/
+        """
+        cls.settings_dir = cls.get_app_attr("settings_dir") or settings_dir
+        cls.subcommand = cls.get_app_attr("subcommand") or subcommand
+
+        # todo(yfried): consider simply searching for option_name instead of
+        # "options"
+        for option_tree in utils.search_tree("options", spec):
+            for option_name, option_dict in option_tree.iteritems():
+                if issubclass(clg.TYPES.get(option_dict.get("type", object),
+                                            object),
+                              cls) and args[option_name] is None:
+                    args[option_name] = clg.TYPES.get(option_dict["type"])(
+                        required=option_dict.get("required")
+                    )
+                    # todo(yfried): this is a good place to trigger
+                    # args[option_name].resolve_value()
+
+    def resolve_value(self, arg_name, defaults=None):
+        """
+        Resolve the argument value with alternative input origins
+
+        Search order:
+            1. Given value from CLI parsing
+            2. Environment variables:
+                [Aa-Zz] -> [A-Z]
+                "-" -> "_"
+            3. defaults dict
+
+        :param arg_name: argument name from clg
+        :param defaults: dict containing default values.
+        """
+        defaults = defaults or {}
+        self.arg_name = arg_name
+        # override get default from env variables
+        # TODO (aopincar): IR env vars should be more uniques
+        if self.value is None:
+            self.value = os.getenv(self.arg_name.upper().replace("-", "_"))
+        # override get default from conf file
+        if self.value is None:
+            self.value = defaults.get(self.arg_name)
+
+
+class YamlFileArgument(ValueArgument):
     """
-    Creates a dictionary based on given configuration file.
+    YAML file input argument.
+    Loads legal YAML from file.
+    Will search for files in the settings directory before trying to resolve
+        absolute path.
+
+    For the argument name is "arg-name" and of subparser "SUBCOMMAND" of
+        application "APP", the default search path would be:
+
+         settings_dir/APP/SUBCOMMAND/arg/name/arg_value
+    """
+
+    def resolve_value(self, arg_name, defaults=None):
+        super(YamlFileArgument, self).resolve_value(arg_name, defaults)
+        search_first = os.path.join(self.get_app_attr("settings_dir"),
+                                    self.get_app_attr("subcommand"),
+                                    *arg_name.split("-"))
+        self.value = find_file(self.value, search_first)
+
+
+def find_file(filename, search_first):
```

Comment:
Maybe move this inside of YamlFileArgument and exetend TopologyArgument from YamlFileArgument. Just an idea.